### PR TITLE
gEDA => Lepton EDA configuration upgrade tools

### PR DESCRIPTION
**Pull request by graahnul-grom**
Extend `liblepton` configuration API, add `Scheme` code and a command-line tool to upgrade legacy `geda*.conf` configuration files using new settings names and produce corresponding `lepton*.conf` files.

New `(conf upgrade)` module, for example, can be used in `lepton-schematic` startup code. It exports the `convert()` function, which is used by new `lepton-upcfg` utility. Passing either `'local`, `'user` or `'system` to that function, one can produce `lepton.conf` (in current directory), `lepton-user.conf` or `lepton-system.conf` configuration files, respectively, with imported and renamed settings.

Mapping between old and new names is stored in the `keys-list` list structure in the `(conf keys)` module.

Simultaneous access to both old and new configuration files is possible because now we have two `SYSTEM` and `USER` configuration contexts in `liblepton`.
The code can switch between them using the `config-set-legacy-mode!` function. So, after the call `(config-set-legacy-mode! #t)`, configuration API will work with `geda*.conf` files, and after `(config-set-legacy-mode! #f)` `lepton*.conf` files will be used.

Example:
```sh
$ ls -l
-rw-r--r--  1 dmn  dmn  246 Jan 24 19:01:51 2019 geda.conf
-rw-r--r--  1 dmn  dmn  970 Jan 21 06:57:09 2019 tb111.sch

$ cat geda.conf
[schematic.gui]
use-docks=false
[gnetlist]
default-net-name=NNN
[gschem]
default-filename=un
[gnetlist.hierarchy]
netname-attribute-separator=__
[export]
margins=10;10;20;20
[gschem.library]
component-attributes=refdes;value
sort=true

$ lepton-upcfg --local
ii: converting [local] configuration...

ii: INPUT: /home/dmn/lepton/pp/tb_111_upcfg/pr/geda.conf

ii: read   [gnetlist]::default-net-name                           = [NNN]
ii: read   [gnetlist.hierarchy]::netname-attribute-separator      = [__]
ii: read   [gschem.library]::component-attributes                 = [(refdes value)]
ii: read   [gschem.library]::sort                                 = [#t]
ii: read   [gschem]::default-filename                             = [un]
ii: read   [export]::margins                                      = [(10 10 20 20)]
ii: read   [schematic.gui]::use-docks                             = [#f]

ii: OUTPUT: /home/dmn/lepton/pp/tb_111_upcfg/pr/lepton.conf

$ cat lepton.conf
[schematic.gui]
use-docks=false

[export]
margins=10;10;20;20;

[schematic]
default-filename=un

[schematic.library]
sort=true
component-attributes=refdes;value;

[netlist.hierarchy]
netname-attribute-separator=__

[netlist]
default-net-name=NNN
```

affects #39

### CODE REVIEW

File: liblepton/scheme/Makefile.am [hunk 1]
```diff
@@ -1,11 +1,22 @@
...
+	geda/attrib.scm \
+	geda/deprecated.scm \
+	geda/os.scm \
+	geda/config.scm \
+	geda/log-rotate.scm \
+	geda/log.scm \
+	geda/library.scm \
+	geda/repl.scm \
+	lepton/version.scm \
+	conf/keys.scm \
```
Comment (vzh) [inline liblepton/scheme/Makefile.am]: I would suggest to rename `conf/keys.scm` to `lepton/legacy-config/keylist.scm`, and `conf/upgrade.scm` to `lepton/legacy-config.scm`, so that the modules were in the `lepton` namespace, and longer, more descriptive names were used for files and procedures.

File: liblepton/scheme/conf/keys.scm [hunk 2]
```diff
@@ -0,0 +1,265 @@
+;; Lepton EDA
+;; liblepton - Lepton's library - Scheme API
+;; Copyright (C) 2019 dmn <graahnul.grom@gmail.com>
+;; License: GPLv2+. See the COPYING file
+;;
+
+( define-module ( conf keys )
```
Comment (vzh) [inline liblepton/scheme/conf/keys.scm]: Please see the above note.

File: liblepton/scheme/conf/keys.scm [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+  #:use-module  ( geda config )
+  #:export      ( conf-keys-list )
```
Comment (vzh) [inline liblepton/scheme/conf/keys.scm]: `config-keylist` would sound better, wouldn't it?

File: liblepton/scheme/conf/keys.scm [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+( define ( conf-keys-list )
+  ; return:
+  keys-list
+)
+
+
+
+; functions to retrieve configuration values:
+;
+( define pfn-get-bool     config-boolean     )
```
Comment (vzh) [inline liblepton/scheme/conf/keys.scm]: Why is this needed?

File: liblepton/scheme/conf/upgrade.scm [hunk 5]
```diff
@@ -0,0 +1,233 @@
+;; Lepton EDA
+;; liblepton - Lepton's library - Scheme API
+;; Copyright (C) 2019 dmn <graahnul.grom@gmail.com>
+;; License: GPLv2+. See the COPYING file
+;;
+
+( define-module ( conf upgrade )
```
Comment (vzh) [inline liblepton/scheme/conf/upgrade.scm]: See the above comment about module naming.

File: liblepton/scheme/conf/upgrade.scm [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+
+  #:use-module  ( ice-9 format )
+  #:use-module  ( geda  config )
+  #:use-module  ( conf  keys   )
+
+  #:export      ( convert )
```
Comment (vzh) [inline liblepton/scheme/conf/upgrade.scm]: I would prefer more descriptive names like, say, `upgrade-config` and avoid using of common names for procedures, since they could clash with imported third party code.

File: liblepton/scheme/conf/upgrade.scm [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+  ( catch #t
+    ( lambda()
+      ( config-load! cfg )
+      ( set! keys (conf-keys-list) )
+      ( set! data
+        ( read-old cfg keys #:report-absent-keys report-absent-keys )
+      )
+    )
+    ( lambda( ex . args )
+      ( throw 'load "Cannot read config:" ex args )
```
Comment (vzh) [inline liblepton/scheme/conf/upgrade.scm]: Will this error be visible in some log files?

File: liblepton/scheme/conf/upgrade.scm [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+  ( data  '() )
+  )
+
+  ;
+  ; read legacy config file (geda*.conf):
+  ;
+
+  ( config-set-legacy-mode! #t ) ; use geda*.conf files
+
+  ( set! cfg ( get-cfg-ctx what ) )
```
Comment (vzh) [inline liblepton/scheme/conf/upgrade.scm]: Defining local variables and re-setting them makes the code less readable in my view. Wouldn't it be better just to use `(let* ...) here?

File: liblepton/scheme/conf/upgrade.scm [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+) ; convert()
+
+
+
+; get configuration context
+;
+; [what]: symbol: 'local, 'user or 'system
+; {ret}:  requested configuration context or #f on failure
+;
+( define ( get-cfg-ctx what )
```
Comment (vzh) [inline liblepton/scheme/conf/upgrade.scm]: Here and below, and in other file code, defining of names after the code they are used in can lead to collisions that would be hard to find, if the names are already defined in some other imported code. I've already stumbled upon such situations in the past. So please avoid this.

File: liblepton/scheme/conf/upgrade.scm [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+  )
+
+  ( for-each
+  ( lambda( entry )
+
+    ( parse-entry entry )
+
+    ( catch #t
+      ( lambda()
+        ( set! val ( pfn cfg grp-old key-old ) ) ; pfn() throws
```
Comment (vzh) [inline liblepton/scheme/conf/upgrade.scm]: One note for future. In Lisp world nobody promises that a sequence of procedures will be evaluated in order. While `guile` docs state it does so, things can be changed in future. Or, say, another Scheme could be used (a few years ago, TinyScheme has been suggested by one of the geda devs). Statements could be processed in parallel then. That is why local variable definition, say, by `let` is better than `set!`. Thus, the functional style of programming, when all procedure arguments do not depend upon others, is better. This is why, AIUI, the folded definitions are used so often. Results of low level functions are directly passed to top level ones. Lisp is not C, please remember this ;-)

File: liblepton/scheme/conf/upgrade.scm [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+;
+( define ( write-new cfg res )
+( let
+  (
+  ( grp-new #f )
+  ( key-new #f )
+  ( val     #f )
+  )
+
+  ( define ( parse-entry e )
```
Comment (vzh) [inline liblepton/scheme/conf/upgrade.scm]: Please consider using of `match` from the `(ice-9 match)` module for such parse functions. There are examples in `lepton-netlist` code.

File: utils/upcfg/lepton-upcfg.in [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+      ( format #t "xx: convert() failed: ['~a]~%  ~a~%" ex args )
+    )
+  )
+
+) ; let
+) ; main()
+
+
+
+( define ( usage exit-code )
```
Comment (vzh) [inline utils/upcfg/lepton-upcfg.in]: The same note as above about definition after usage.

File: liblepton/scheme/conf/upgrade.scm (hunk 8 above)
Comment (graahnul-grom) [reply liblepton/scheme/conf/upgrade.scm]: This function is called in two places to get different contexts. Similar assignments maintains the code symmetry, so to say, and makes it easier to understand, IMHO. :)

File: liblepton/scheme/conf/upgrade.scm (hunk 9 above)
Comment (graahnul-grom) [reply liblepton/scheme/conf/upgrade.scm]: Thank you for the advice.
Code rearranged.

File: liblepton/scheme/conf/upgrade.scm (hunk 11 above)
Comment (graahnul-grom) [reply liblepton/scheme/conf/upgrade.scm]: Didn't know about `match()`. :-) It is really helpful function, thank you.

File: liblepton/scheme/conf/upgrade.scm (hunk 10 above)
Comment (graahnul-grom) [reply liblepton/scheme/conf/upgrade.scm]: Using language that cannot provide basic code execution guaranties in any serious project would be disastrous.

File: liblepton/scheme/conf/keys.scm (hunk 4 above)
Comment (vzh) [reply liblepton/scheme/conf/keys.scm]: Hmm, fair enough...

**Comment by vzh** [comment]
OK, now it looks a bit prettier. Thanks!
